### PR TITLE
fix(runtime costs): adjust runtime costs

### PR DESCRIPTION
**Pull request by olonho**
Executed our runtime under emulator as described in
https://github.com/nearprotocol/nearcore/blob/master/runtime/runtime-params-estimator/emu-cost/README.md
to figure out the exact machine (x86 in our case) instruction counter required to perform basic operations.
Then normalized so that SHA256 computations cost aproximately the same.

Comparision with the old costs:
base: time=126224222 insn=88429085 insn/time=0.70
read_memory_base: time=1629369577 insn=773923915 insn/time=0.47
read_memory_byte: time=123816 insn=1266318 insn/time=10.23
write_memory_base: time=76445225 insn=95327640 insn/time=1.25
write_memory_byte: time=809907 insn=874985 insn/time=1.08
read_register_base: time=639340699 insn=743687111 insn/time=1.16
read_register_byte: time=63637 insn=31955 insn/time=0.50
write_register_base: time=0 insn=73022018 insn/time=n/a
write_register_byte: time=0 insn=7 insn/time=n/a
utf8_decoding_base: time=0 insn=162997218 insn/time=n/a
utf8_decoding_byte: time=591904 insn=96359 insn/time=0.16
utf16_decoding_base: time=0 insn=323612618 insn/time=n/a
utf16_decoding_byte: time=9095538 insn=6592872 insn/time=0.72
sha256_base: time=710092630 insn=651203831 insn/time=0.92
sha256_byte: time=5536829 insn=6626443 insn/time=1.20
keccak256_base: time=710092630 insn=1097138631 insn/time=1.55
keccak256_byte: time=5536829 insn=5700785 insn/time=1.03
keccak512_base: time=710092630 insn=1074127478 insn/time=1.51
keccak512_byte: time=5536829 insn=10762570 insn/time=1.94
log_base: time=0 insn=0 insn/time=n/a
log_byte: time=0 insn=0 insn/time=n/a
storage_write_base: time=21058769282 insn=13359081673 insn/time=0.63
storage_write_key_byte: time=23447086 insn=19416882 insn/time=0.83
storage_write_value_byte: time=9437547 insn=7084061 insn/time=0.75
storage_write_evicted_byte: time=0 insn=0 insn/time=n/a
storage_read_base: time=19352220621 insn=9743366720 insn/time=0.50
storage_read_key_byte: time=4792496 insn=7180693 insn/time=1.50
storage_read_value_byte: time=139743 insn=47159 insn/time=0.34
storage_remove_base: time=109578968621 insn=11051260769 insn/time=0.10
storage_remove_key_byte: time=9512022 insn=9337676 insn/time=0.98
storage_remove_ret_value_byte: time=0 insn=94568 insn/time=n/a
storage_has_key_base: time=20019912030 insn=9586751818 insn/time=0.48
storage_has_key_byte: time=4647597 insn=7175135 insn/time=1.54
storage_iter_create_prefix_base: time=28443562030 insn=0 insn/time=0.00
storage_iter_create_prefix_byte: time=442354 insn=0 insn/time=0.00
storage_iter_create_range_base: time=25804628282 insn=0 insn/time=0.00
storage_iter_create_from_byte: time=429608 insn=0 insn/time=0.00
storage_iter_create_to_byte: time=1302886 insn=0 insn/time=0.00
storage_iter_next_base: time=24213271567 insn=0 insn/time=0.00
storage_iter_next_key_byte: time=0 insn=0 insn/time=n/a
storage_iter_next_value_byte: time=1343211668 insn=0 insn/time=0.00
touching_trie_node: time=1 insn=1 insn/time=1.00
promise_and_base: time=0 insn=0 insn/time=n/a
promise_and_per_promise: time=672136 insn=541550 insn/time=0.81
promise_return: time=34854215 insn=97824365 insn/time=2.81

Test plan
---------
cargo test

### CODE REVIEW

**Comment by nearmax** [comment]
https://github.com/nearprotocol/nearcore/blob/3cab57953dbaf1f89c5f6026aa121ed519b50c73/runtime/near-runtime-fees/src/lib.rs#L115 needs to also be updated

**Comment by nearmax** [comment]
> log_base: time=0 insn=0 insn/time=n/a
> log_byte: time=0 insn=0 insn/time=n/a

This is weird. With timing it is clear why this is near 0 seconds because it is almost instantaneous, but it should've accounted for some number of instructions..

Let's copy the values of `utf8_decoding_base, utf8_decoding_byte` here just to stay on the safe side.

**Comment by nearmax** [comment]
> write_register_base: time=0 insn=73022018 insn/time=n/a
> write_register_byte: time=0 insn=7 insn/time=n/a

This is great! Finally we have a tool sensitive enough to measure it.

**Comment by nearmax** [comment]
@bowenwang1996 Could you provide code pointers to @olonho since he is new to the process?

**Comment by bowenwang1996** [comment]
Yeah. Follow the examples in `scripts/migrates` and write a script that changes config as you did in this PR. Also bump `PROTOCOL_VERSION` in `core/chain-configs/src/lib.rs`

**Comment by olonho** [comment]
Implemented migration tools/

**Comment by nearmax** [comment]
@olonho `RuntimeFeesConfig` is still needed. @bowenwang1996 is it actually the right way to implement migration tool? AFAIU it consumes the old config and spits out the new one, so it should be also overwriting the values of the fees. Correct?

File: scripts/migrations/11-gas-price-fix.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+home = sys.argv[1]
+output_home = sys.argv[2]
+
+config = json.load(open(os.path.join(home, 'output.json')), object_pairs_hook=OrderedDict)
+
+assert config['protocol_version'] == 10
+
+config['protocol_version'] = 11
+
```
Comment (bowenwang1996) [inline scripts/migrations/11-gas-price-fix.py:19]: You need to overwrite the old values. See how it is done for example in `migrate/scripts/6-state-stake.py`

**Comment by olonho** [comment]
> @olonho I've fixed the migration script. Please let me know if I should merge this.

Thanks a lot! Please don’t merge it so far, I have to decide with @nearmax if changes in https://github.com/nearprotocol/nearcore/pull/2621 makes sense, and then update costs with the new data.

**Comment by olonho** [comment]
Adjusted costs per new computatins.

File: scripts/migrations/11-runtime-cost-adjustment.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        "utf8_decoding_byte": 289342653,
+        "utf16_decoding_base": 3593689800,
+        "utf16_decoding_byte": 167519322,
+        "sha256_base": 4530046500,
+        "sha256_byte": 24116301,
+        "keccak256_base": 5867223186,
+        "keccak256_byte": 21469644,
+        "keccak512_base": 5798128650,
+        "keccak512_byte": 36651981,
+        "log_base": 0,
```
Comment (nearmax) [inline scripts/migrations/11-runtime-cost-adjustment.py:130]: It looks like these were not updated using https://github.com/nearprotocol/nearcore/pull/2621 . Log should become non-zero.

File: scripts/migrations/11-runtime-cost-adjustment.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        "storage_has_key_byte": 28376217,
+        "storage_iter_create_prefix_base": 0,
+        "storage_iter_create_prefix_byte": 0,
+        "storage_iter_create_range_base": 0,
+        "storage_iter_create_from_byte": 0,
+        "storage_iter_create_to_byte": 0,
+        "storage_iter_next_base": 0,
+        "storage_iter_next_key_byte": 0,
+        "storage_iter_next_value_byte": 0,
+        "touching_trie_node": 1,
```
Comment (nearmax) [inline scripts/migrations/11-runtime-cost-adjustment.py:152]: Could we make `touching_trie_node` temporarily equal to `storage_read_base`/log_16(20000) ?
Until we do certain optimizations to our trie this operation will have non-zero cost, so for the safety sake we could at least approximate it. The above formula would approximate it.

File: scripts/migrations/11-runtime-cost-adjustment.py (hunk 2 above)
Comment (olonho) [reply scripts/migrations/11-runtime-cost-adjustment.py:130]: Right, they were just hardcoded to be zeros, will fix.

File: scripts/migrations/11-runtime-cost-adjustment.py (hunk 3 above)
Comment (olonho) [reply scripts/migrations/11-runtime-cost-adjustment.py:152]: Fixed. Both fixes added to https://github.com/nearprotocol/nearcore/pull/2621.

File: core/chain-configs/src/lib.rs [hunk 4]
```diff
@@ -7,4 +7,4 @@ pub use genesis_config::{
 };
 
 /// Current latest version of the protocol
-pub const PROTOCOL_VERSION: u32 = 10;
+pub const PROTOCOL_VERSION: u32 = 11;
```
Comment (frol) [inline core/chain-configs/src/lib.rs:10]: @bowenwang1996 we should make sure we don't collide with #2504. Just a heads up.

Comment (bowenwang1996) [reply core/chain-configs/src/lib.rs:10]: We will not because the state migration check on CI will fail.

**Pull request merged**